tests/system/Router/RouteCollectionTest.php

```diff
@@ -337,6 +337,104 @@ static function ($routes): void {
         $this->assertSame($expected, $routes->getRoutes());
     }
 
+    public function testGroupNestedWithOuterOptionsWithoutInnerOptions(): void
+    {
+        $routes = $this->getCollector();
+
+        $routes->group(
+            'admin',
+            ['namespace' => 'Admin', 'filter' => ['csrf']],
+            static function ($routes) {
+                $routes->get('dashboard', static function () {
+                });
+
+                $routes->group('profile', static function ($routes) {
+                    $routes->get('/', static function () {
+                    });
+                });
+            }
+        );
+
+        $expected = [
+            'admin/dashboard' => [
+                'namespace' => 'Admin',
+                'filter'    => ['csrf'],
+            ],
+            'admin/profile' => [
+                'namespace' => 'Admin',
+                'filter'    => ['csrf'],
+            ],
+        ];
+        $this->assertSame($expected, $routes->getRoutesOptions());
+    }
+
+    public function testGroupNestedWithOuterAndInnerOption(): void
+    {
+        $routes = $this->getCollector();
+
+        $routes->group(
+            'admin',
+            ['filter' => ['csrf']],
+            static function ($routes) {
+                $routes->get('dashboard', static function () {
+                });
+
+                $routes->group(
+                    'profile',
+                    ['filter' => ['honeypot']],
+                    static function ($routes) {
+                        $routes->get('/', static function () {
+                        });
+                    }
+                );
+            }
+        );
+
+        $expected = [
+            'admin/dashboard' => [
+                'filter' => ['csrf'],
+            ],
+            'admin/profile' => [
+                'filter' => ['honeypot'],
+            ],
+        ];
```

Comment:
This is still confusing to me but I guess it's because we can have only one filter when `Config\Feature::multipleFilters` is disabled.

Will this gonna end up as:

```php
$expected = [
    'admin/dashboard' => [
        'filter' => ['csrf'],
    ],
    'admin/profile' => [
        'filter' => ['csrf', 'honeypot'],
    ],
];
```
when `multipleFilters` option will be enabled?

Reply:
No. it is just because using `array_merge()`.
https://3v4l.org/tskmm

It is the current behavior that filters are not merged, and also the reason why the Note in the user guide was added.

One possible behavior would be to merge filters as well, but then it would not be possible to delete filters in inner group. Multiple filters are already the default in 4.5.

Reply:
Good point about removing filters. 

I don't know... now I'm leaning towards staying with the current behavior and not inheriting anything. Because the new behavior seems even more confusing - it's inherited only if there are no other filters. I can imagine people adding a new filter over time and "losing" functionality.

Thanks to the current note, at least we have clear rules to follow.

Reply:
The current behavior is not that straightforward either.
The current behavior is that if you don't pass an options array to the inner group, the outer options are inherited.

That's why when you set outer filter and add namespace option to inner group, the filter that set in outer suddenly will be gone for the inner route.

It seems that this issue is not so easy to conclude.

Reply:
Yes, this is a bit more complex than I thought at first. Especially if we add namespaces to it.

Personally, I would probably merge all the declared filters from the external to the internal group using `array_merge_recursive` and introduce something like `filterExclude` to remove the filters if needed. However, this seems a bit too complex... not sure if we want to make such big changes.

 Although the rules would be clear and need no further explanation.

Reply:
> No. it is just because using `array_merge()`. https://3v4l.org/tskmm
> 
> It is the current behavior that filters are not merged, and also the reason why the Note in the user guide was added.
> 
> One possible behavior would be to merge filters as well, but then it would not be possible to delete filters in inner group. Multiple filters are already the default in 4.5.

https://3v4l.org/2jijf
While we are working with options as an array, why not combine the outer with the inner?

I believe the group should be inheritable internally. Including namespace. To exclude filters, it is possible to create an additional key or, as before, configure it in `Config/Filters`

Reply:
I'm leaning towards "all or none". Either we should completely merge all keys, including filters as @neznaika0 suggests - or we should exclude filters from any inheritance and require them to be redefined in subroutes. I think the former solution is more intuitive but the latter solution is more flexible. 

Given that Filters are a significant nexus for security concerns I think this is important to consider carefully.

Reply:
I also am leaning towards "all or none". 

But if you specify namespaces, merging does not make sense.
See https://3v4l.org/ifMuX
When merging all, namespaces must be overwritten and filters must be merged.

If we go with none, the filter that was previously working, specified only for the outer group, will no longer work on the inner routes.

Reply:
That's right, namespace needs to be replaced.

Reply:
@neznaika0 `$multipleFilters` was already removed in 4.5. Do you want to revert it?

Reply:
I still go back to the beginning and `array_merge()` seems to be the way to go.
At least the implementation will be the simplest.

Reply:
> @neznaika0 `$multipleFilters`was already removed in 4.5. Do you want to revert it?

How do we make exceptions when merging filters? Just ignore it? 

For example, we don't need an 'outer_filter' in the group - `['auth', 'outer_filter'] >> ['auth','inner_filter']`
Add except to `Config`? If "Yes", it looks fine.

```php
// Routes
$routes->group('', ['filter' => 'outer_filter', 'inner_filter'], function ($routes) {
    $routes->get('route1', function () {});
    $routes->get('route2', function () {});
});

// Filters
public array $globals = [
        'before' => [
            'inner_filter' => ['except' => ['route1/*', 'route2/*'],
        ],
        'after' => [
            'inner_filter' => ['except' => ['route1/*', 'route2/*'],
        ],
    ];
```
Or back, add only 100% common filters in the group. Individual set up inside.

```php
// Routes
$routes->group('', ['filter' => 'outer_filter'], function ($routes) {
    $routes->get('route1', function () {}, ['filter' => 'inner_filter']);
    $routes->get('route2', function () {}, ['filter' => 'inner_filter']);
});
```

Reply:
`array_merge()` replaces all filters in outer group with filters in inner group.
See the test:
https://github.com/codeigniter4/CodeIgniter4/pull/7981/files#diff-9e755caf334ca07345b43d4fdc3986ac1f9bb9594c6f78ebdec59bc780699b73R371-R400

Reply:
I looked, in my opinion it's wrong. We can't fine-tune the filters any more. The same example with authorization - we want to have it for many routes, and have an additional filter inside.

Reply:
3. We want to have 1-2 outer filters and 0-5 inner.

Okay. I understand your position. If it's not necessary, I don't say anymore

Reply:
Sorry, I don't get what you say. What is 1-2 outer and 0-5 inner?

Do you want to put a filter in outer group and remove it in inner group?

Reply:
> > @neznaika0 `$multipleFilters`was already removed in 4.5. Do you want to revert it?
> 
> How do we make exceptions when merging filters? Just ignore it?
> 
> For example, we don't need an 'outer_filter' in the group - `['auth', 'outer_filter'] >> ['auth','inner_filter']` Add except to `Config`? If "Yes", it looks fine.
> 
> ```
> // Routes
> $routes->group('', ['filter' => 'outer_filter', 'inner_filter'], function ($routes) {
>     $routes->get('route1', function () {});
>     $routes->get('route2', function () {});
> });
> 
> // Filters
> public array $globals = [
>         'before' => [
>             'outer_filter' => ['except' => ['route1/*', 'route2/*'],
>         ],
>         'after' => [
>             'outer_filter' => ['except' => ['route1/*', 'route2/*'],
>         ],
>     ];
> ```
> 
> Or back, add only 100% common filters in the group. Individual set up inside.
> 
> ```
> // Routes
> $routes->group('', ['filter' => 'outer_filter'], function ($routes) {
>     $routes->get('route1', function () {}, ['filter' => 'inner_filter']);
>     $routes->get('route2', function () {}, ['filter' => 'inner_filter']);
> });
> ```
> 
> EDIT: to `outer_filter` except

Numbers - count filters. ['filter' => '1', '2', '3'] 
0 - without filters

Reply:
> Why? You can control filters 100%. The rules are:
> 
> 1. outer filters are applied if you don't specify filter at all in inner group.
> 
> 2. if you want to change filters in inner group, specify all filters.

I think this is the best we can do for now, and as @kenjis noted is also the simplest implementation.

Reply:
@neznaika0 Sorry, I don't get what you mean at all.

Reply:
What exactly? A message with numbers? Or "how to merge filters"?

The figures are an example of the number of filters used: from 1 to X pieces. I want to show that it is convenient to have 2 filters outside and add (or not) 5 inside. Then the routes will have from 2 to 7 filters.

In your version, it is necessary to write 2-7 filters for each router. 
 `['filter' => 'outer_filter1', 'outer_filter2', 'inner_filter1',  ..., 'inner_filter5',]`

The main purpose of the question -`array_merge_recursive($params['outer']['filter'], $params['inner']['filter'])`

Reply:
Thank you for explanation. I got what you say.

Yes, the options are
1. use `array_merge()` for filters
2. use `array_merge_recursive()` for filters

When you have 2 filters outside and add (or not) 5 inside, in my version,
yes, you need to specify 7 filters when you add filters in inner routes:
```php
['filter' => ['outer_filter1', 'outer_filter2', 'inner_filter1',  ..., 'inner_filter5']]
```
But none (not specify `filter` key in the option array) when you don't add filters, because the outer 2 filters will be merged.

Using `array_merge_recursive()` seems better if you never remove outer filters in inner routes.
But if you want to remove a outer filter in inner routes, there is no way.

Reply:
I am very sorry for my translation. 

Yes, to exclude the outside filter, I gave an example of `except` in the **Config**. Or we need to write such a filter for all routes except unnecessary ones (this is your option)

Reply:
I disagree with the new feature to exclude route filters in the `Config\Filter`.
The `Config\Filter` and route filters are the same filters, but I think mixing these settings is a source of confusion.

Also, I don't like to add a new key for excluding filters in the route option.
It also is complicated.

Reply:
Well, this does not cancel your method - we leave the outside filters empty and configure each route individually

Reply:
`Config\Filter` defines filters for URI paths, not for routes.
`except` excepts the filter to the URI path in `Config\Filter`.
https://codeigniter4.github.io/CodeIgniter4/incoming/filters.html#except-for-a-few-uris

Filters for a route are different things. If you specify a filter to a route, it is always applied.
The definitions in `Config\Filter` does nothing to route filters.

What I don't like is to mix the two kind of filters. That is, `except` in `Config\Filter` excepts route filters, too.
It is really complicated and confusing.

Reply:
@michalsn @neznaika0 @MGatner I changed my mind. I sent another PR #8033 that behaves like michalsn expects in https://github.com/codeigniter4/CodeIgniter4/pull/7981#discussion_r1336782260